[Backend] Implement layout conversion within warps with shuffle idx

Mogball opened the pull request:
Warp shuffles are affine transformations on the lane ID. This PR implements an approach that recognizes we can represent affine transformations by augmenting the linear layout with a dummy dimension and then decompose the mapping from src to dst layout into a series of affine transformations that represent warp shuffles.

It turns out for all distributed layouts, we can implement this transformation using a single index shuffle (mask + permute).

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -657,6 +664,165 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+  // because `A` and `B` are distributed layouts (subpermutation matrices), we
+  // know the transformation from `A` to `B` is invariant to the register ID.
+  // Reduce `C` to a transformation from source lane to destination lane.
+  assert(C.sublayoutIsZero(kLane, kRegister) &&
+         C.sublayoutIsZero(kRegister, kLane) &&
+         "expected a subpermutation matrix");
```
This is not true. You can have a layout with 5 lanes and 5 registers per lane that simply "transposes the data", meaning that `lane_i` gets the info of `reg_i` and the other way around.

If you have a fast path when the condition above is true, then this would be a very neat optimisation on top tho.

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -657,6 +664,165 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+  //
+  // Note that this gives `l' = A * l + c` when the dummy dimension is set to 1.
+  //
+  // Butterfly shuffles are not affine because 0 ⊕ 1 = 1, thus zero input
```
nit "are not linear"

lezcano commented:
FWIW, It might be a bit too much for a first PR, but something that'd be a nice north start would be to discover the neat algorithm in `convertMMAV3To8BitsDotOperand` automatically.

Mogball commented:
I went a bit overboard with the implementation instead of keeping it simple (couldn't resist)... but the solution is much more general now. I'm not sure the PR is in a landable state however, given the complexity. I also really need to set this down to push on other things but I would love comments on the approach when you guys have time.

> FWIW, It might be a bit too much for a first PR, but something that'd be a nice north start would be to discover the neat algorithm in `convertMMAV3To8BitsDotOperand` automatically.

The approach I found doesn't quite do that yet, because it doesn't account for bytepacking the i32s for warp shuffles and there isn't TargetInfo interface to the byte permute instructions (the ROCDL dialect doesn't expose the AMD one, but we can probably do something similar as the `nvgpu` dialect if necessary), but it does provide a general method for decomposing the layout conversion into  (register permute, warp shuffle, register permute). It just needs more constraints to ensure the permutation also puts the right sub-i32 values next to each other and use byte permutes instead of selects when possible.

Jokeren commented:
Will take a look soon. Thanks!
I implemented something similar to this approach initially and found a certain "search" process is required. The code was not clean and also cannot handle the 8-bit scenario, so I gave it up.

lezcano commented:
So cool! Will look into it tomorrow!

lezcano commented on include/triton/Tools/LinearLayout.h:
```diff
@@ -534,6 +534,15 @@ class LinearLayout {
     return reshapeOuts({{*getOutDimNames().begin(), getTotalOutDimSize()}});
   }
 
+  // Concatenates two layouts by their input dimensions. The layouts must have
+  // the same output dimensions and sizes and different input dimensions. The
+  // input dimensions of this layout are placed before those of 'other'.
+  [[nodiscard]] LinearLayout concatIns(const LinearLayout &other) const;
+  // Concatenates two layouts by their output dimensions. The layouts must have
+  // the same input dimensions and sizes and different output dimensions. The
+  // output dimensions of this layout are placed before those of 'other'.
+  [[nodiscard]] LinearLayout concatOuts(const LinearLayout &other) const;
```
how are these different from `operator*`?

Mogball replied:
concatIns does something like 

```
A.concatIns(B) ->
[ A ]
[ B ]

A.concatOuts(B) -> 
[ A | B ]

A * B -> 
[ A ? ]
[ ?  B ]
```

I'm not entirely sure what `operator*` does, tbh, but it is different than these two.

lezcano replied:
Nit. I believe in your example the matrices are flipped (`A.concatIns(B) -> [A | B]`).

Right, writing proper documentation for `operator*`  is in my TODO list. Note that `A * B` does what you described if `A` and `B`  have disjoint ins and outs. If `A` and `B` have the same outs and disjoint ins, `A * B = [A | B]`. Similarly if they have disjoint outs but same ins.

I might be missing something here, but could you check whether `operator*` works for your use case? If not, perhaps providing a specific example in which they differ and what's the result of each of them would help solving this mistery

Mogball replied:
Perhaps I didn't explain it properly in the documentation of the functions, but supposed I have

```
// Identity mapping from register->register
 - register=1 -> (1)
   register=2 -> (2)
   register=4 -> (4)
   register=8 -> (8)
   register=16 -> (16)
where out dims are: [register (size 32)]

// Zeros mapping from register->lane
 - register=1 -> (0)
   register=2 -> (0)
   register=4 -> (0)
   register=8 -> (0)
   register=16 -> (0)
where out dims are: [lane (size 32)]
```

I want to construct 

```
 - register=1 -> (1, 0)
   register=2 -> (2, 0)
   register=4 -> (4, 0)
   register=8 -> (8, 0)
   register=16 -> (16, 0)
where out dims are: [lane (size 32)]
```

But `operator*` gives

```
 - register=1 -> (1, 0)
   register=2 -> (2, 0)
   register=4 -> (4, 0)
   register=8 -> (8, 0)
   register=16 -> (16, 0)
   register=32 -> (0, 0)
   register=64 -> (0, 0)
   register=128 -> (0, 0)
   register=256 -> (0, 0)
   register=512 -> (0, 0)
where out dims are: [register (size 32), lane (size 32)]
```

Mogball replied:
I guess I could have used `LinearLayout::identity1D(32, kRegister, kRegister) * LinearLayout::zeros1D(1, kRegister, kLane, 32)` but sometimes I'm trying to build a layout from less trivial quadrants. I.e. imagine I want to splice two layouts together `A(reg, lane) -> (reg, lane)` and `B(reg, lane) -> (reg, lane)` such that the resulting layout is

```
[ A.sublayout(reg, reg)   A.sublayout(reg, lane) ]
[ B.sublayout(lane, reg)  B.sublayout(lane, lane) ]
```

lezcano replied:
That's fair enough. Do you mind leaving a comment saying that this can be thought of as the "inverse" of `sublayout` (feel free to explain this idea better lol)

lezcano commented on include/triton/Tools/LinearLayout.h:
```diff
@@ -616,6 +625,11 @@ class LinearLayout {
   // the same, and are the bases the identity?
   bool squareSublayoutIsIdentity(ArrayRef<StringAttr> dimNames) const;
 
+  // Is the sublayout defined from dimNames to dimNames a subpermutation matrix?
+  // I.e. the layout matrix is formed by selecting unique rows from the identity
+  // matrix and adding zero rows.
+  bool squareSublayoutIsPermutation(ArrayRef<StringAttr> dimNames) const;
+
```
Can we please create a file with helper functions? Either in this PR or a different one. We are just dumping every function as a method with looong names, and it's starting to look a bit too much like Java :P

Mogball replied:
Yeah let me do that now. I can factor a bunch of stuff into there. Where should I put this file?

lezcano replied:
(althought if the diff is small that might be alright)

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -655,6 +665,87 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+  // selects.
+  LinearLayout reducedP2 = P2inv.removeZeroBasesAlongDim(kLane);
+  SmallVector<Value> results(shflOuts.size());
+  for (int i = 0, e = results.size(); i != e; ++i) {
```
why this format??

Mogball replied:
Old habits die hard. This is from the upstream MLIR coding style guides. The idea is that if you re-query the size in the condition, it's indication that you expect the size to change, e.g.

```C++
for (auto i = 0; i < astNodes.size(); ++i) {
  processMoreNodesRecursively(astNodes[i], &astNodes);
}
```

I'm not really sure what our coding style is other than a mishmash of the styles of other MLIR codebases I've seen lol

Jokeren commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +329,172 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  // around (plus broadcasting, if necessary).
+  //
+  // Warp shuffles allow indexing into another lane, but does not allowing
+  // selecting the register. Suppose we decompose `C` into `C = P2 ∘ W ∘ P1`,
```
Why it's `C = P2 ∘ W ∘ P1` but not `C = P1 ∘ W ∘ P2`?

Jokeren commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +329,172 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  //
+  // We can see that P' selects rows (i.e. bases) from the upper half (register)
+  // and combines them with the lower half (lane). Because the goal is to select
+  // register bases `i` where C(r,l)[i] != 0, we know P'*C(r,r) = 0. Note that
```
Is it `P'*C(r,r) = 0` or `P'*C(r,l) = 0`?

Jokeren commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +329,172 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  // `W^-1 ∘ C'` will be
+  //
+  //   [ I R ] [ C'(r,r) C'(r,l) ] = [ ... C'(r,l) + R*C'(l,l) ]
+  //   [ 0 L ] [ C'(l,r) C'(l,l) ] = [ ... L*C'(l,l)           ]
```
Similarly, I don't understand why W^-1 = 

```
[ I R ]
[0 L]
```

But let's assume it's correct

Jokeren commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +329,172 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  //   W^-1 ∘ C' = [ ... 0 ]
+  //               [ ... 1 ]
+  //
+  // Thus L = C'(l,l)^-1, and A = -C'(r,l) * C'(l,l)^-1. (0 - LL) = LL in GF(2).
```
What is `A` here?

Mogball replied:
Should be R. Let me fix that

Jokeren commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +329,172 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  // Since P1 cannot change lanes, we know that
+  //
+  //   W^-1 ∘ C' = [ ... 0 ]
+  //               [ ... 1 ]
```
1-> identity? So it's `I` instead of `1`?

Jokeren reviewed (commented):
Finished the first round. I get the high level idea, which is similar to the gather function. But I do need a second round review to dig into math details. Right now, seems like there are a few symbols that I misunderstood.

lezcano commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +329,172 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  //
+  // P2 can only change the register mapping within a thread. Constrain P2 as:
+  //
+  //   P2(x) = [ I P ] [ reg  ] = [ reg + P(lane)  ]
```
+ in `GF(2)` is just the `xor`.

Jokeren replied:
Oh, I see. This is the `⊕` operator that Justin defined

Mogball reviewed (commented):
@Jokeren thanks for the review! And yeah it did end up looking similar to gather in the way the selects are handled

Mogball commented:
For the most part, the implementation seems correct, but there are a few conversions of 4d+ tensors that are broken. I'm digging into those but it's not clear how the rank affects the code here

Jokeren commented:
I also noticed the time spent in Python tests has been increased? Can you please investigate a bit?

Mogball commented:
> I also noticed the time spent in Python tests has been increased? Can you please investigate a bit?

Yeah thanks for pointing that out. I haven't measured perf yet, but I think there are many cases where this is slower than shared memory (vector stores, stmatrix, etc.)

lezcano commented:
now WTF!
I pressed the wrong key on my keyboard and it closed this PR, and since it automatically deleted the branch, I cannot reopen it.

Can you push again that branch and reopen?

Mogball commented:
> now WTF! I pressed the wrong key on my keyboard and it closed this PR, and since it automatically deleted the branch, I cannot reopen it.
> 
> Can you push again that branch and reopen?

are you using the "New GitHub Merge Experience^TM"?

Mogball commented:
Please fill out this field.

Mogball commented:
> I also noticed the time spent in Python tests has been increased? Can you please investigate a bit?

It turns out that for particularly uncooperative layouts, a truly large amount of selects get emitted and LLVM dies.

Mogball commented:
Okay, CI is green and the python test times are unchanged. I will validate performance before I move this out of draft. Also, this PR has fixes for bugs that are unrelated to the PR itself that I will separate out.

Mogball commented:
I've verified that there aren't any performance regressions. Interestingly, I cannot seem to find any microbenchmarks where warp shuffles perform any different than shared memory conversions. I'm playing around with kernels that look like

```python
@triton.jit
def kernel(a_ptr, b_ptr):
    idx = (tl.arange(0, 16)[None, :] + tl.arange(0, 16)[:, None])
    x = tl.load(a_ptr + idx)
    for i in range(10000000):
        x += tl.trans(x.reshape((2, 2, 2, 2, 2, 2, 2, 2)), (1, 3, 5, 7, 0, 2, 4, 6)).reshape(16, 16)
    tl.store(b_ptr + idx, x)
```

Which AFAICT just runs layout conversion in a loop. The only observable difference in the SASS between the smem and warp shuffle versions (besides the implementation of the conversion) is that the SMEM version tends to get unrolled by a larger factor (4 or 8) probably since it tends to use fewer instructions. Maybe I'm measuring the wrong things!

At the very least, this reduces the amount of shared memory that has to be allocated for kernels without changing the performance 🙃 .

Jokeren commented:
Thanks for the update. Will take a pass tomorrow or Sunday!

Jokeren commented on lib/Dialect/TritonGPU/IR/Dialect.cpp:
```diff
@@ -1441,7 +1441,7 @@ SmallVector<unsigned> basesPerDim(const LinearLayout::BasesT &namedBases,
   SmallVector<unsigned> ret(rank, 1);
   auto nonZero = [](auto val) { return val != 0; };
-  int nonZeroIdx = -1;
+  int nonZeroIdx = 0;
```
Can you rebase on top of main?

Mogball replied:
Yes

Jokeren commented on lib/Analysis/Utility.cpp:
```diff
@@ -328,6 +329,217 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  for (int i : llvm::seq(C.getInDimSizeLog2(kLane))) {
+    ArrayRef<int32_t> /*C(l,(r,l))[i]*/ lowerHalfRow = C.getBasis(kLane, i);
+    assert(lowerHalfRow.size() == 2);
+    if (/*C(l,r)*/ lowerHalfRow[0] != 0) {
```
```suggestion
    if (/*C(l,r)[i]*/ lowerHalfRow[0] != 0) {
```

Jokeren commented on lib/Analysis/Utility.cpp:
```diff
@@ -328,6 +329,217 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  }
+
+  // Find rows in the upperhalf that can be selected by P' to make the lane
+  // components in the lower half non-zero.
```
I don't understand why it's called upper half and lower half. Isn't it left and right?

Jokeren replied:
I was referring to

```
  //   [ I  0 ] [ C(r,r) C(r,l) ] = [ C(r,r)             C(r,l)           ]
  //   [ P' I ] [ C(l,r) C(l,l) ]   [ P'*C(r,r)+C(l,r)   P'*C(r,l)+C(l,l) ]
```

Jokeren replied:
Oh, I get it. Never mind. It is the lower and upper from the perspective of `C`

Jokeren commented on include/triton/Analysis/Utility.h:
```diff
@@ -168,6 +168,24 @@ class GatherLoweringHelper {
+// Nearly all layout conversions that only require data movement within a warp
+// can be implemented this way.
+struct DecomposedWarpConversion {
+  triton::LinearLayout P1, W, P2;
```
So this is `P2inv` instead of `P2`?

Mogball replied:
Yes. I'll rename it and clarify why the inverse is returned.

Jokeren commented on test/Conversion/cvt_to_llvm.mlir:
```diff
@@ -0,0 +1,153 @@
+  // 6         T10:0  T26:0   T11:0  T27:0
+  // 7         T14:0  T30:0   T15:0  T31:0
+  //
+  // Where the registers change every 2 lanes like [0, 4, 1, 5, 2, 6, 3, 7].
```
Every 2 lanes? or 4 lanes?

Mogball replied:
Every 2 lanes, since it wraps around after lane 16

peterbell10 commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -308,8 +308,11 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+        transferWithinWarp(op, *decomposedCvt, adaptor, rewriter);
+        return success();
+      }
       return transferWithinBlock(op, srcLayout, dstLayout, adaptor, rewriter);
```
Random thought. Since data is only transferred within a warp, you could also try still using smem but changing `barrier` to `bar.warp.sync` between writing and reading. This should improve latency when the warps issue the barrier on different cycles. (Though we would still need a barrier between `ConvertLayoutOp`s).

Jokeren replied:
Sure, we can definitely have a test on it.

Mogball replied:
That's a good idea!

Jokeren replied:
@Mogball Any speedups you have see for this particular optimization?

Mogball replied:
I haven't tried this one yet. @lezcano also mentioned that he was planning to look into a more optimized shmem implementation at some point, based on @apgoucher's algorithm.

lezcano replied:
Yep, but that one is orthogonal to everything here really

Mogball replied:
I left a TODO in the code to try this later

Mogball commented on lib/Analysis/Utility.cpp:
```diff
@@ -328,6 +329,217 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  // Don't use this codegen approach for small element sizes, since due to
+  // upcasting to i32, it can be less efficient than storing to shared memory
+  // until shuffling packed values is implemented.
+  if (!isa<PointerType>(srcTy.getElementType()) &&
```
This only handles element bitwidths greater than or equal to 32, including pointers. I'll clarify the comment. Can you elaborate on why this might result in no measured speedups?

lezcano replied:
In case you want to improve perf for small types, you can pack them. The maximum packing is given by (two to the power of) `dst[kRegister].intersection(src[kRegister])`

Jokeren replied:
> Can you elaborate on why this might result in no measured speedups?

I only meant that we didn't handle small data types so we didn't see significant speedups for corresponding test cases.

Mogball replied:
Yeah makes sense. I added handling for small data types when there isn't much vectorization, since I observed 1.5-2x speedups for 16 and 8 bit types in a few cases

Jokeren replied:
That's nice. Can you do me a favor to sort out all layout conversions that you observed speedups? I'll use it for paper writing. Thanks!

Mogball commented on include/triton/Tools/LinearLayout.h:
```diff
@@ -344,8 +344,8 @@ class LinearLayout {
-  static LinearLayout zeros1D(int32_t size, StringAttr inDim,
-                              StringAttr outDim);
+  static LinearLayout zeros1D(int32_t size, StringAttr inDim, StringAttr outDim,
+                              int32_t outDimSize = 1);
```
I clarified what this does in the comment. It creates a non-surjective layout if `outDim` has a size greater than 1, since the layout will only map to 0. Let me know if you think this is too specific to be part of this function API.

Mogball commented on include/triton/Tools/LayoutUtils.h:
```diff
@@ -9,6 +9,12 @@ namespace mlir::triton {
+
+// Is the sublayout defined from dimNames to dimNames a subpermutation matrix?
+// I.e. the layout matrix is formed by selecting unique rows from the identity
+// matrix and adding zero rows.
```
I think it should be rows. A zero row would mean that varying an input bit does not change the output bits.

Jokeren replied:
Oh sorry for confusion. From the perspective of `f2reduce`, I think columns are inputs and rows are outputs.

Mogball commented on test/Conversion/tritongpu_to_llvm.mlir:
```diff
@@ -771,29 +771,29 @@ module attributes {"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 4 : i32} {
   // CHECK: llvm.mlir.global external @global_smem
   // CHECK-LABEL: convert_layout_blocked_blocked
-  tt.func @convert_layout_blocked_blocked(%arg0: tensor<16x16xf32, #blocked0>) {
+  tt.func @convert_layout_blocked_blocked(%arg0: tensor<32x32xf32, #blocked0>) {
```
Yes

Mogball commented on lib/Analysis/Utility.cpp:
```diff
@@ -328,6 +329,217 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  // P2 can only change the register mapping within a thread. Constrain P2 as:
+  //
+  //   P2 = [ I P ]
+  //        [ P I ]
```
Yes thanks for catching this.

Mogball commented:
@Jokeren I double-checked my benchmark script and realized I made an incredibly silly mistake. Now I am actually measuring differences between the two implementations. In some cases, conversion using warp shuffles is up to 2x faster, and in some cases, it is 2x slower! I think this has to do with the number of selects that get emitted and the smem iterations / vector length. I'm trying to narrow down a basic heuristic that can be used to make a good choice.

Mogball commented:
@Jokeren Here's a hopefully relatively detailed description of the layout conversions where speedups were observed. First, here's the benchmark script:

```python
import triton
import pathlib
import torch
import triton.language as tl


@triton.jit
def kernel1(a_ptr, b_ptr):
    idx = (tl.arange(0, 16)[None, :] + tl.arange(0, 16)[:, None])
    x = tl.load(a_ptr + idx)
    for i in range(1000):
        x += x
    tl.store(b_ptr + idx, x)


TSIZE=32
dtype='i64'
torch_dtype=torch.int64

ttgir = f"""
#blocked = #ttg.blocked<{{sizePerThread = [32, 1], threadsPerWarp = [1, 32], warpsPerCTA = [1, 1], order = [1, 0]}}>
#test = #ttg.blocked<{{sizePerThread = [16, 2], threadsPerWarp = [2, 16], warpsPerCTA = [1, 1], order = [1, 0]}}>
module attributes {{"ttg.num-ctas" = 1 : i32, "ttg.num-warps" = 1 : i32, ttg.target = "cuda:90", "ttg.threads-per-warp" = 32 : i32}} {{
  tt.func public @kernel(%arg0: !tt.ptr<{dtype}> {{tt.divisibility = 16 : i32}}, %arg1: !tt.ptr<{dtype}> {{tt.divisibility = 16 : i32}}) attributes {{noinline = false}} {{
    %c1_i32 = arith.constant 1 : i32
    %c1000000_i32 = arith.constant 100000 : i32
    %c0_i32 = arith.constant 0 : i32
    %0 = tt.make_range {{end = {TSIZE} : i32, start = 0 : i32}} : tensor<{TSIZE}xi32, #ttg.slice<{{dim = 0, parent = #blocked}}>>
    %1 = tt.expand_dims %0 {{axis = 0 : i32}} : tensor<{TSIZE}xi32, #ttg.slice<{{dim = 0, parent = #blocked}}>> -> tensor<1x{TSIZE}xi32, #blocked>
    %2 = tt.make_range {{end = {TSIZE} : i32, start = 0 : i32}} : tensor<{TSIZE}xi32, #ttg.slice<{{dim = 1, parent = #blocked}}>>
    %3 = tt.expand_dims %2 {{axis = 1 : i32}} : tensor<{TSIZE}xi32, #ttg.slice<{{dim = 1, parent = #blocked}}>> -> tensor<{TSIZE}x1xi32, #blocked>
    %4 = tt.broadcast %1 : tensor<1x{TSIZE}xi32, #blocked> -> tensor<{TSIZE}x{TSIZE}xi32, #blocked>
    %5 = tt.broadcast %3 : tensor<{TSIZE}x1xi32, #blocked> -> tensor<{TSIZE}x{TSIZE}xi32, #blocked>
    %6 = arith.addi %4, %5 : tensor<{TSIZE}x{TSIZE}xi32, #blocked>
    %7 = tt.splat %arg0 : !tt.ptr<{dtype}> -> tensor<{TSIZE}x{TSIZE}x!tt.ptr<{dtype}>, #blocked>
    %8 = tt.addptr %7, %6 : tensor<{TSIZE}x{TSIZE}x!tt.ptr<{dtype}>, #blocked>, tensor<{TSIZE}x{TSIZE}xi32, #blocked>
    %9 = tt.load %8 : tensor<{TSIZE}x{TSIZE}x!tt.ptr<{dtype}>, #blocked>

    %10 = scf.for %arg2 = %c0_i32 to %c1000000_i32 step %c1_i32 iter_args(%arg3 = %9) -> (tensor<{TSIZE}x{TSIZE}x{dtype}, #blocked>)  : i32 {{
      %x = ttg.convert_layout %arg3 : tensor<{TSIZE}x{TSIZE}x{dtype}, #blocked> -> tensor<{TSIZE}x{TSIZE}x{dtype}, #test>

      %res = arith.addi %x, %x : tensor<{TSIZE}x{TSIZE}x{dtype}, #test>
      %13 = ttg.convert_layout %res : tensor<{TSIZE}x{TSIZE}x{dtype}, #test> -> tensor<{TSIZE}x{TSIZE}x{dtype}, #blocked>
      scf.yield %13 : tensor<{TSIZE}x{TSIZE}x{dtype}, #blocked>
    }}
    %11 = tt.splat %arg1 : !tt.ptr<{dtype}> -> tensor<{TSIZE}x{TSIZE}x!tt.ptr<{dtype}>, #blocked>
    %12 = tt.addptr %11, %6 : tensor<{TSIZE}x{TSIZE}x!tt.ptr<{dtype}>, #blocked>, tensor<{TSIZE}x{TSIZE}xi32, #blocked>
    tt.store %12, %10 : tensor<{TSIZE}x{TSIZE}x!tt.ptr<{dtype}>, #blocked>
    tt.return
  }}
}}
"""

temp_file = pathlib.Path("test.ttgir")
temp_file.write_text(ttgir)
kernel = triton.compile(str(temp_file))


@triton.testing.perf_report(
    triton.testing.Benchmark(
        x_names=['size'],  # Argument names to use as an x-axis for the plot.
        x_vals=[1],  # Different possible values for `x_name`.
        x_log=True,  # x axis is logarithmic.
        line_arg='provider',  # Argument name whose value corresponds to a different line in the plot.
        line_vals=['triton'],  # Possible values for `line_arg`.
        line_names=['Triton'],  # Label name for the lines.
        styles=[('blue', '-')],  # Line styles.
        ylabel='GB/s',  # Label name for the y-axis.
        plot_name='cvt',
        args={},  # Values for function arguments not in `x_names` and `y_name`.
    ))
def benchmark(size, provider):
    a = torch.arange(TSIZE * TSIZE, device="cuda", dtype=torch_dtype).reshape(TSIZE, TSIZE)
    b = torch.empty_like(a)
    #kernel[(1, )](a, b, num_warps=1)

    quantiles = [0.5, 0.2, 0.8]
    ms, min_ms, max_ms = triton.testing.do_bench(lambda: kernel[(1, 1, 1)](a, b), quantiles=quantiles)
    return ms, min_ms, max_ms


a = torch.arange(TSIZE * TSIZE, device="cuda", dtype=torch_dtype).reshape(TSIZE, TSIZE)
b = torch.empty_like(a)
print(kernel.asm["llir"])

benchmark.run(print_data=True)
```

`num_warps` is locked to 1 for all the tests. If you consider blocked layouts, I tested layout conversions that altered `threadsPerWarp = [a, b]` into `threadsPerWarp = [a*c, b/c]` where `c = [1/32, 1/16, ..., 8, 16, 32]` as appropriate based on `a` and `b`, for a variety of `sizePerThread` and column-major vs row-major.

The closer `c` is to 1, the more iterations the shmem conversion requires, but the fewer selects need to get emitted for warp shuffles. For `c = 2,1/2`, warp shuffles are always faster, but for `c = 4,1/4`, warp shuffles are typically slower, and are sometimes the same speed depending on other factors.

I also tested against vectorized shmem when `c = 2,1/2`. Here's the data I collected https://docs.google.com/spreadsheets/d/1JV1D7PZHlkVvms_MtEwNRhfLaINZIfiMkRaMVv0hpgY/edit?usp=sharing

I also tested layouts where `iterations=1` and no selects are emitted for warp shuffles, and warp shuffles are still slightly faster. I don't remember what layout exactly triggers this case, but I think it's the one in the benchmark script since those were the last layouts I was testing.

Jokeren commented:
@Mogball thanks for the results. IIUC, packing i8/i16 was not implemented for the evaluation? It's good enough to demonstrate the effectiveness already, but please let me know if packing is on your roadmap.

lezcano reviewed (commented):
At last I found time to read through this! I got the main idea, but I have not reviewed the whole code, as I think we can do better. I'll leave it up to you to decide whether you wan to re-work this PR, or whether you want to reimplement the generic algo. If it's the former, I am happy to do a detailed review of the whole thing.

Let's consider the running example from the comments to make things clearer
```
src = {register = [[1,0], [2,0]], lane = [[0,1], [0,2]]}
dst = {register = [[0,1], [2,0]], lane = [[1,0], [0,2]]}
```

So, we have two linear functions, S, T: H -> M (Source and Target from Hardware into a Matrix). In the example, M = F_2^2 x F_2^2 and H is "the same as M but with labels" (i.e. isomorphic to M). Let's assume there is no thread broadcasting in S or T for now for simplicity.

First, we want to find a subspace of M so that the elements it defines belong to different threads in S and T. A way to do this is:
- Find a basis of the intersection of I = S[lane] \cap T[lane]. In our example it's {(0, 2)}
- Consider C = {x ^ y | (x, y) \in zip(S[lane]\I, T[lane]\I)}. In our example it's {(1, 1)}.
A basis of a subspace with the property above is given by B = C \cup I (check that indeed the elements {(0, 0), (0, 2), (1, 1), (1, 3)} all belong different threads).

Then we complete this basis into a basis of the whole space. In our example, we choose R = {(0, 1), (2, 0)} (we could have chosen (1, 0) as the first element of the basis, it doesn't matter).
R gives us the iterations of our algorithm. In the first iteration we'll interchange the elements of B, in the next one we'll interchange the elements of B ^ (0, 1), then B ^ (2, 0) and finally B ^ (2, 1). In each iteration we'll want to get the target values for the elements of these sets, so we'll need to compute the relevant sources that own them.

Maths tells us that this split of M will have all the threads in every iteration, so it leads to feasible warp shuffles. Even more, we see that in this example S^{-1}(x) = T^{-1}(x) for x \in B and for x \in B ^ (2, 0), which means that we don't need to emit those shuffles because the elements are already in the right position (this can be computed looking at the kernel of src^{-1} o dst - id where src and dst are the matrices with the zeros removed (probably, I haven't fully checked)).

Mogball commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +331,230 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  //
+  // Warp shuffles allow indexing into another lane, but does not allowing
+  // selecting the register. Suppose we decompose `C` into `C = P1 ∘ W ∘ P2`,
+  // where `W` is a warp shuffle and `P1` and `P2` are (lane-dependent) register
```
I defined a single warp shuffle as

```
W = [ I R' ]
    [ 0 L  ]
```

Which is just one index shuffle.

Mogball commented on include/triton/Analysis/Allocation.h:
```diff
@@ -58,6 +58,12 @@ struct ScratchConfig {
   }
 };
 
+// For a layout conversion between `srcTy` and `dstTy`, return the vector length
```
Can you elaborate on that proper vectorization means?

lezcano replied:
I just meant choosing the maximum packing we can do per transaction looking at the size of S[thread] \cup T[thread]

Mogball commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +331,230 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+  //  2          T0:2   T0:3   T2:2   T2:3
+  //  3          T1:2   T1:3   T3:2   T3:3
+  //
+  // Note for each destination register, two lanes want two different registers
+  // in the same source lane (T0:0 -> T0:0, T1:0 -> T0:1). This is impossible to
+  // represent with a warp shuffle, because the source lane (e.g. T0) can only
+  // supply one of its registers as the shuffle value.
```
Yes, but this algorithm finds a register permutation so that it can be done in just 1 warp shuffle. This algorithm breaks down the conversion transformation into `register permute -> one index shuffle -> register permute`.

Jokeren replied:
I'm a bit confused. I think @lezcano meant that the entire conversion can be done in two shuffles? But there @Mogball 's algorithm actually should use four shuffles? Please correct me if I were wrong

Mogball replied:
Yes I misunderstood. Mario's algorithm does the whole thing in 2 shuffles total. I was referring to 1 shuffle "per tensor element in each thread". We chatted quickly about this offline.

Jokeren replied:
That makes sense. IMO it's fine to merge this PR first and improve it later.

Mogball commented:
@lezcano Thanks for the detailed explanation! One of the sketchiest parts of the current implementation is the part where I manually scan the bases in the `cvt` layout and rearrange them to determine the first register permutation. I might try to rework this incorporating some of your ideas.

The main difference between the two algorithms is the one I implemented always generates 1 warp shuffle since it has a rigid structure, whereas your algorithm may generate more due to breaking down the transformation into iterations. The algorithm I implemented only fails when there is lane broadcasting, because it makes some of the sublayouts noninvertible. I put off dealing with this in the future.

I actually think that there are cases where generating more warp shuffles is better than trying to permute registers to generate 1, because the register permutation can have a significant runtime cost. I will try to think about how to find a decent composition of the two algorithms.

peterbell10 commented:
> I actually think that there are cases where generating more warp shuffles is better than trying to permute registers to generate 1, because the register permutation can have a significant runtime cost.

FYI based on [this table](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#arithmetic-instructions) I think a shuffle may cost the same as 4 moves, assuming mov and fma have similar throughput.

Mogball commented:
@lezcano After thinking about this more, I am not certain that either algorithm is more general, since they both (as written) handle the same cases, but generate different code! In brief, the approach I implemented always tries to decompose the conversion into `reg prmt + warp shuffle + reg prmt`, whereas your approach would break the conversion into a series of warp shuffles.

However, in order to handle broadcasting in the conversion, I am pretty sure we will need to codegen at least 2 warp shuffles instead of 1. If the number of shuffles required to handle broadcasting is variable (instead of just one index plus one butterfly shuffle), then that is where I think your approach might be better suited. But broadcasting isn't needed to support the current use cases that matter.

The other factor is that sometimes it is better to emit multiple warp shuffles rather than try to emit a single shuffle with register permutations. But for smaller data types, register permutations become cheaper because we can use byte permute instructions, so the benefits aren't clear at the moment.

I think a possible next step would be to implement your algorithm separately and try to understand the cases in which it generates better code, and maybe figure out how to unify the approaches.

Mogball commented:
> > I actually think that there are cases where generating more warp shuffles is better than trying to permute registers to generate 1, because the register permutation can have a significant runtime cost.
> 
> FYI based on [this table](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#arithmetic-instructions) I think a shuffle may cost the same as 4 moves, assuming mov and fma have similar throughput.

That's good to know. It would certainly help build a cost model for an algorithm that has to pick between generating register permutations and warp shuffles.

Mogball commented:
Update: I chatted with Mario and I did misunderstand a few of the details.  In particular, for the example

```
            lane
register       0      1      2      3
 0          T0:0   T0:1   T2:0   T2:1  
 1          T1:0   T1:1   T3:0   T3:1  
 2          T0:2   T0:3   T2:2   T2:3  
 3          T1:2   T1:3   T3:2   T3:3  
```

Since some of the elements are already in the right spot (identity from dst->src), it's possible to shuffle just a subset of the data. In particular, 

```
            lane
register       0      1      2      3
 0        ✅T0:0 1️⃣T0:1 ✅T2:0 1️⃣T2:1  
 1        1️⃣T1:0 ✅T1:1 1️⃣T3:0 ✅T3:1  
 2        ✅T0:2 2️⃣T0:3 ✅T2:2 2️⃣T2:3  
 3        2️⃣T1:2 ✅T1:3 2️⃣T3:2 ✅T3:3  
```

One can see that only two shuffles are needed to put the data in the right places.

lezcano commented on lib/Analysis/Utility.cpp:
```diff
@@ -329,6 +331,230 @@ unsigned ScanLoweringHelper::getScratchSizeInBytes() {
+    return {};
+
+  // HACK: Workaround AMD codegen path generating transient invalid layouts.
+  auto isInvalidDotEnc = [](RankedTensorType type) {
+    auto dotEnc = dyn_cast<DotOperandEncodingAttr>(type.getEncoding());
+    return dotEnc && dotEnc.getKWidth() == 0;
+  };
+  if (isInvalidDotEnc(srcTy) || isInvalidDotEnc(dstTy))
+    return {};
```
This is fine, but you can be more explicit by seeing if the parent is not an NvidiaMmaLayout (or a Blocked layout... does the FMA layout work?).
cc @antiagainst for the AMD skips

Mogball replied:
Added a check for AMDMfma/Wmma attr

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -642,6 +655,77 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+  StringAttr kRegister = str_attr("register");
+  StringAttr kLane = str_attr("lane");
+  assert(!cvtNeedsSharedMemory(op.getSrc().getType(), op.getType()));
+  auto [P1, Cp, P2inv, reducedP1, reducedP2] = std::move(decomposed);
```
Oh, I didn't know you could unpack PODs like this! This is very nice!

lezcano commented on lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp:
```diff
@@ -642,6 +655,77 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+    assert(out.size() == 1);
+    Value srcRegIdx = out.front().second;
+    // The size of the input lane dimension is the number of selects to emit.
+    // TODO(jeff): For dtypes smaller than i32, we can use byte permutes and
+    // shuffle multiple values at a time.
```
I think so! I want to get to a point where we have a better/more general implementation of reg to reg permutations that uses prmt and here we just call that and everything just works.

lezcano reviewed (approved):
Are there python tests that exercise the correctness of this path? It might be good to add a few pairs of layouts that do so to `convert2d` or so.

Regarding the algorithm proposed in the previous review, we'll implement it in a different PR.

Mogball commented:
@lezcano thanks for the review!

> Are there python tests that exercise the correctness of this path? It might be good to add a few pairs of layouts that do so to convert2d or so.

There are, but I'll explicitly add a few.
